### ISSUE TITLE
Strange results from fp16 benchmarks on Alexnet on CIFAR10

### ISSUE THREAD

**Issue by OneRaynyDay** [apache/incubator-mxnet#11842]
## Description
The benchmark code is [here](https://github.com/OneRaynyDay/mxnet-fp16-benchmark).
For some reason, performing inference on fp16 models does not actually create any speed ups for me:

Forward pass on the entire CIFAR10 dataset with DataLoader on fp16:
57.90255570411682 seconds

Forward pass on the entire CIFAR10 dataset with DataLoader on fp32:
48.82793569564819 seconds

Why is fp16 taking even longer than fp32? I perform a cast on the network after creating it, and cast the data before inputting it according to the tutorial [here](https://mxnet.incubator.apache.org/faq/float16.html).

@rahul003 

## Environment info (Required)

Running on a p3.8xlarge

```
nvidia-smi:

+-----------------------------------------------------------------------------+
| NVIDIA-SMI 384.130                Driver Version: 384.130                   |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla V100-SXM2...  Off  | 00000000:00:1B.0 Off |                    0 |
| N/A   46C    P0    53W / 300W |   5100MiB / 16152MiB |      6%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla V100-SXM2...  Off  | 00000000:00:1C.0 Off |                    0 |
| N/A   42C    P0    34W / 300W |     10MiB / 16152MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   2  Tesla V100-SXM2...  Off  | 00000000:00:1D.0 Off |                    0 |
| N/A   39C    P0    37W / 300W |     10MiB / 16152MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   3  Tesla V100-SXM2...  Off  | 00000000:00:1E.0 Off |                    0 |
| N/A   40C    P0    37W / 300W |     10MiB / 16152MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0     44367      C   /opt/conda/envs/user/bin/python             2152MiB |
|    0     56344      C   /opt/conda/envs/user/bin/python              906MiB |
|    0     61965      C   /opt/conda/envs/user/bin/python             1188MiB |
|    0     79156      C   /opt/conda/envs/user/bin/python              844MiB |
+-----------------------------------------------------------------------------+
```

```
(dockerized)[ray_zhang@i-0555fc08518f7d737 xxx]$ pip freeze | grep mx
You are using pip version 9.0.1, however version 10.0.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
keras-mxnet==2.2.0
mxnet-cu90==1.3.0b20180715
```

```
(dockerized)[ray_zhang@i-0555fc08518f7d737 xxx]$ python --version
Python 3.6.3 :: Intel Corporation
```

```
(dockerized)[ray_zhang@i-0555fc08518f7d737 xxx]$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 16.04.4 LTS
Release:	16.04
Codename:	xenial
```

**Comment by nswamy** [comment]
@rahul003

**Comment by rahul003** [comment]
Hi Ray,
Cifar doesn't lend itself well for fp16 computation. 

Nvidia Tensor Cores essentially perform the computation D = A * B + C, where A and B are half precision matrices, while C and D could be either half precision or full precision. The tensor cores are most efficient when dimensions of these matrices are multiples of 8. This means that Tensor Cores can not be used in all cases for fast float16 computation. When training Cifar10 dataset, the tensors involved are sometimes smaller, and Tensor Cores can not always be used. The computation in that case falls back to slower algorithms and using float16 turns out to be slower than float32 on a single GPU. Note that when using multiple GPUs, using float16 can still be faster than float32 because of reduction in communication costs.

**Comment by rahul003** [comment]
Please review the performance section of the tutorial you [linked](https://mxnet.incubator.apache.org/faq/float16.html) for more tips on getting good performance when using Fp16.

**Comment by nswamy** [comment]
@OneRaynyDay Please close the issue if this @rahul003 answered your question.

**Comment by safrooze** [comment]
@rahul003 Have you confirmed if this is a performance issue? If so, can you please remove Question label and add Performance label?

**Comment by rahul003** [comment]
For CIFAR I had observed similar performance. 

Here's a related thread with my observations and logs of nvprof. 
https://github.com/apache/incubator-mxnet/issues/9774#issuecomment-366835890
I was told here that the dimensions of convolutions for CIFAR are not suitable for tensor cores. https://github.com/apache/incubator-mxnet/issues/9774#issuecomment-367057821

I'm not sure if we can improve fp16 support for CIFAR. @OneRaynyDay mentioned that he's transforming the image size to 224x224, in which case I don't think why he doesn't see any speedup. You can verify with synthetic data of 224x224 size to see that it does improve the speed. Need more profiling to understand what's going on here. 

I request a committer to add Performance label to the issue.

**Comment by thomelane** [comment]
@mklissa what's been your experience of using FP16 with CIFAR-10? did you see any speed up?

**Comment by mklissa** [comment]
In my case there was definitely a speed-up with FP16 on CIFAR10. It is important to add the flag "multi_precision=True" to the optimizer as FP16 might create issues with the gradients.

**Comment by rahul003** [comment]
@mklissa Did you notice a speedup with a single GPU as well? Maybe it has improved now. 
With multi GPU I did see performance improvement because of reduced communication.

**Comment by mklissa** [comment]
Hi Rahul,
I have seen speed-ups on single GPU as well (around 40%), as well as multi GPU.

**Comment by rahul003** [comment]
Great!

**Comment by piyushghai** [comment]
@OneRaynyDay Did the addition of the flag multi_precision=True to the optimizer help in achieving a speed up ? 
@sandeep-krishnamurthy Can you please close this issue as it looks like the commenters on this thread have seen perf boosts using clever flag tricks.
